sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -514,3 +514,178 @@ case class StringToMap(text: Expression, pairDelim: Expression, keyValueDelim: E
+      InternalRow.fromSeq(newValues)
+    }
+  }
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val structGen = struct.genCode(ctx)
+    val addOrReplaceFieldsGens = valExprs.map(_.genCode(ctx))
+    val resultCode: String = {
+      val structVar = structGen.value
+      type NullCheck = String
+      type NonNullValue = String
+      val existingFieldsCode: Seq[(FieldName, (NullCheck, NonNullValue))] =
+        ogStructType.fields.zipWithIndex.map {
+          case (structField, i) =>
+            val nullCheck = s"$structVar.isNullAt($i)"
+            val nonNullValue = CodeGenerator.getValue(structVar, structField.dataType, i.toString)
+            (structField.name, (nullCheck, nonNullValue))
+        }
+      val addOrReplaceFieldsCode: Seq[(FieldName, (NullCheck, NonNullValue))] =
+        fieldNames.zip(addOrReplaceFieldsGens).map {
+          case (fieldName, fieldExprCode) =>
+            val nullCheck = fieldExprCode.isNull.code
+            val nonNullValue = fieldExprCode.value.code
+            (fieldName, (nullCheck, nonNullValue))
+        }
+      val newFieldsCode = loop(existingFieldsCode, addOrReplaceFieldsCode)
+      val rowClass = classOf[GenericInternalRow].getName
+      val rowValuesVar = ctx.freshName("rowValues")
+      val populateRowValuesVar = newFieldsCode.zipWithIndex.map {
+        case ((_, (nullCheck, nonNullValue)), i) =>
+          s"""
+             |if ($nullCheck) {
+             | $rowValuesVar[$i] = null;
+             |} else {
+             | $rowValuesVar[$i] = $nonNullValue;
+             |}""".stripMargin
+      }.mkString("\n|")
+
+      s"""
+         |Object[] $rowValuesVar = new Object[${dataType.length}];
+         |
+         |${addOrReplaceFieldsGens.map(_.code).mkString("\n")}
+         |$populateRowValuesVar
+         |
+         |${ev.value} = new $rowClass($rowValuesVar);
+          """.stripMargin
+    }
+
+    if (nullable) {
+      val nullSafeEval =
+        structGen.code + ctx.nullSafeExec(struct.nullable, structGen.isNull) {
+          s"""
+             |${ev.isNull} = false; // resultCode could change nullability.
+             |$resultCode
+             |""".stripMargin
+        }
+
+      ev.copy(code =
+        code"""
+          boolean ${ev.isNull} = true;
+          ${CodeGenerator.javaType(dataType)} ${ev.value} = ${CodeGenerator.defaultValue(dataType)};
+          $nullSafeEval
+          """)
+    } else {
+      ev.copy(code =
+        code"""
+          ${structGen.code}
+          ${CodeGenerator.javaType(dataType)} ${ev.value} = ${CodeGenerator.defaultValue(dataType)};
+          $resultCode
+          """, isNull = FalseLiteral)
+    }
+  }
+
+  override def prettyName: String = "add_fields"
+
+  private type FieldName = String
+
+  /**
+   * Recursively loop through addOrReplaceFields, adding or replacing fields by FieldName.
+   */
+  @scala.annotation.tailrec
```

Comment:
Could you rewrite this code with a loop? https://github.com/databricks/scala-style-guide#recursion-and-tail-recursion

Reply:
replaced it with a `foldLeft` loop, thanks